### ISSUE TITLE
How can this work well with wasm?

### ISSUE THREAD

**Issue by dtribble** [tc39/proposal-weakrefs#16]
Wasm is starting to need weak references.  This issue is for discussion of how this proposal can work there.

**Comment by xtuc** [comment]
Sorrry for my ignorance, but why WebAssembly needs weakrefs. Could you please link to some discussion?

**Comment by lars-t-hansen** [comment]
WebAssembly wants weak references initially for its [host-binding proposal](https://github.com/WebAssembly/host-bindings).  There is a strawman [here](https://github.com/WebAssembly/host-bindings/issues/14), the writeup contains a number of motivating details.  It was written in part because TC39 has been dragging its (admittedly very many) feet on weakrefs and similar functionality :)

Since this thread is about how well the current proposal works with wasm, I'll provide some initial reactions now that I have read through the proposal (but only done so once, so really inviting a discussion here to correct my misinterpretations, as much as anything).

By and large, and except for the next point below, I think wasm would be satisfied with the current weakref proposal for its host-binding needs; I don't know what we'll need for support for GC-in-wasm in the future, but I assume we'll need more than this.   I don't think this proposal will be in the way, though, and maybe that will have limited exposure to the host language anyway.

The biggest red flag with this proposal for wasm right now is the lack of within-turn weakref resolution, aka #11.  By not having this you force a wasm application that has its own event loop (this will likely be common for programs that have anything resembling threads which, Spectre notwithstanding, we expect to happen) to poll the GC by: setting up a run-me-again-soon event; returning to the event loop, hoping that some finalizer jobs are maybe waiting so that finalizers may be run (but hopefully not too many!), and then waiting for the run-me-again-soon event to fire so that we can do something more.  This plays poorly with runloops that wait on mutexes and so on, of course it can be done but it strikes me as both messy and expensive.

In general I am skeptical about how "the system" decides when to run finalizers and how many of them to run, because this leads both to brittleness (essentially threading issues) and jankiness that the user program will have to work hard to control (essentially critical sections + queues of pending finalizers that it manages itself).  I suspect a "finalizer pump" system (getNextExecutorOrNull() coupled with some type of event that signals the pump is primed again) would provide both better control and within-turn weakref resolution, though I'm not saying the ergonomics of that will necessarily be fantastic either.

**Comment by dtribble** [comment]
@xtuc Mostly what @lars-t-hansen said.  At the recent TC39, several people expressed interest in that connection.

**Comment by dtribble** [comment]
@lars-t-hansen The ES weak reference proposal is quite motivated by precisely the use case you mention in your proposal, and a strong customer may be the thing to get it moving forward. Thus, I want to make sure that it addresses your requirements.  

An aside on weak reference model: in the general case, the "guardians" approach suffers the same resurrection and layered collection issues as other pre-morten approaches. It works OK for the proxy-facade case, because there is a flat layer of facades connecting to wasm reference but breaks if e.g., the "facades" have a tree structure of their own (e.g., it was a tree of DOM-like nodes referencing corresponding WASM objects). These issues are mentioned in the proposal; I'm happy to discuss them further.

In your scenarios, does WASM ever have strong pointers back into ES objects? (and thus potentially at the "facade" objects)? I would expect that it does for many applications.

The ES proposal is to specify how weak references and finalization appear in ES.  It did not expose interfaces for the policy choices that an ES runtime might make.  For exposing weak references to wasm, it may be appropriate to expose additional hooks.

Having the JS runtime schedule finalization is a convenience (and performance/reliability improvement for most code).  It was designed to coexist and operate correctly with the following complementary "finalization":
- a comm system keeps a collection of WeakRefs (in order to map from remote/wasm IDs back to the facade-proxy)
- at any time, it can scan that collection and detect any reference that was already collected (without waiting for finalization), and proactively "finalize". 

Extending the `RemoteConnection` example from the proposal:`
```js
  // finalize any dropped references
  sendDrops() {
    this.remotes.forEach((weakref, remoteId) => {
      if (!weakref.deref()) {
        this.dropRef(remoteId);
      }
    }
  }```
 
We can consider various options to allow platform assistance to the above conceptual algorithm, since the platform can arrange to have an easier time finding already-collected references. An approach like this was used in the original post-mortem solution in Smalltalk: each weak array internally knew it's collected indices, and so a client could iterate through them on demand.  Note that this patterns does NOT introduce new timing exposure.  The GC algorithm is unchanged; a "weak" reference that has been dereferenced in a turn will be strong until the end of the turn, so this won't observe any additional GC timing.

**Comment by lars-t-hansen** [comment]
Wasm will almost certainly have strong pointers back into the ES heap.  Specifically, there's talk of an 'anyref' type in the host bindings proposal and my understanding is that these anyrefs could be stored in wasm "tables", which are effectively global variables.  (Along those lines, wasm already has strong pointers back into the ES heap from both tables and export maps, but largely to benign objects such as functions and WebAssembly.Global objects.)  So yes, we should assume that wasm can reference these JS facade objects in some way.

I've noted elsewhere (#20) that the `get()` making the pointer strong for the remainder of the turn is a problem for in-turn weakref processing, this is not a deep point of course, just mentioning it again.

**Comment by erights** [comment]
Hi @lars-t-hansen ,
Since we're talking about wasm as opposed to wasm-gc, the only use for wasm observable gc-weakness seems to be for the wasm-js boundary. What is a concrete realistic usecase of a wasm program that both
   * does not return to the JS event loop, and
   * does need observable gc weakness for issues at the wasm-js boundary?

**Comment by lars-t-hansen** [comment]
@erights, that's a good observation.

To recap, as we've discussed in the past, we believe a common mode for wasm program will be "threaded", where threads which are not the main thread will be running in a `while (e = waitForSharedMemoryEvent()) doEvent(e)` loop.  Empirically such a loop can respond to events much more quickly than a normal event loop can manage events.  Initially, as for JS, the "threads" will be workers, but since wasm can share compiled code across workers they will be much cheaper to create than JS workers that would reinstantiate the program in each worker.  The asm.js/wasm runtime for C/C++ already spaws workers like these to handle various tasks so as not to block the main thread.  These threads frequently rely on the main thread for interaction with the user and OS when that can't be handled in the thread, passing data back and forth through shared memory.

In this case, the thread will never return to its proper event loop, as it has no reason to do so.  Of course, this program may not have very much JS either, so how important really is the weakref use case here?

We don't truly know.  We just know that it is almost certain that each of these workers will have some JS to handle tasks that cannot be expressed in wasm but that can be handled locally by means of callouts, and it is plausible that some wasm objects (addresses) will become represented by JS objects, and that those objects might be weakly-held facades/proxies.  When that happens, we will want to avoid having to return to the true event loop in the hope that some weakrefs might be collectible.  Instead, some callout from wasm to JS would trigger any finalization that is pending.

I know you asked for "concrete" and "realistic" and that the above is not quite that, but it's what I have at the moment.

**Comment by flagxor** [comment]
Key point here is that not yielding to the event loop until termination is the normal pattern when emscripten emulates pthreads.
Thus, if a pthread based wasm program makes any kind of use of an API surface implemented in another wasm module (imagine a WebSQL or filesystem API polyfill), there will be no opportunity for finalization notifications.

**Comment by erights** [comment]
Hi @flagxor I get that. But recall that all JS is per-wasm-thread (i.e., per JS agent), so the wasm-loop vs JS-event-loop dichotomy is aligned with differential uses of the wasm-JS host boundary. While it is certainly possible for a non-yielding wasm-loop to make use of its own thread-local JS across a host boundary for which observable gc-weakness may be useful, I am still looking for a usecase that is concrete and compelling.

Until then, my intuition is that a reasonable embedding of a multi-threaded wasm program into JS is that there is one JS agent that is responsible for most of the interaction between that wasm program and the JS world, and that this wasm-thread, however minor or major its role in this wasm-program, does yield to its JS event loop.

By contrast, any wasm-thread that does not yield can make no use of any of the other browser[*] functionality that is already tied to the JS event loop: rendering, ui events, network events, promises, async functions, ... . Because of these limitations, ***any*** wasm thread that does not yield to the event loop of its JS agent is already limited to using a tiny fraction of the functionality of JS-in-browser. Given this long list of functionality it already cannot use in this limited context, I don't why it would be especially burdensome to add observable gc-weakness to this list of limitations.

[*] A similar though less familiar list applies to JS-in-node.

**Comment by dtribble** [comment]
I wrote up some potential enhancements or changes to the proposal that also address some of these concerns in the Wiki:  

https://github.com/tc39/proposal-weakrefs/wiki/Support-for-long-wasm-jobs
https://github.com/tc39/proposal-weakrefs/wiki/WeakRefGroups  and

**Comment by lars-t-hansen** [comment]
Dean, I can't quite tell if you're in favor of the WeakRefGroups or not.  I also don't know if they would be useful (even though they seem well-motivated).

I'm not convinced yet by the proposals for the long-running wasm jobs since they seem like they have applicability that is extremely narrow.  I understand their appeal in the case of ensuring that no weakrefs have been deref'd from JS in the turn so that we can allow the turn counter to advance without danger, but the restrictions they impose seem severe right now.

In both cases I think we won't get much further without experimenting with a prototype in the context of real code.

**Comment by bmeck** [comment]
forgot to cross ref this a while ago : https://github.com/tc39/proposal-weakrefs/issues/28

**Comment by littledan** [comment]
I believe the cleanupSome method should meet the need for long Wasm jobs. Please open another issue if you have more specific needs.
